### PR TITLE
QUIC: Delay ack support

### PR DESCRIPTION
**Pull request by scw00**
This pr try to add delay ack feature.

### CODE REVIEW

File: iocore/net/quic/QUICAckFrameCreator.h [hunk 1]
```diff
@@ -23,47 +23,76 @@
...
-  size_t size();
-  void clear();
-  void sort();
-  void forget(QUICPacketNumber largest_acknowledged);
-
-  const std::list<QUICPacketNumber> list() const;
-
-  QUICPacketNumber largest_ack_number();
-  ink_hrtime largest_ack_received_time();
+#define QUIC_ACK_CREATOR_MAX_DELAY 25 * HRTIME_MSECOND
```
Comment (scw00) [inline iocore/net/quic/QUICAckFrameCreator.h]: Should use max_delay_ack params.

File: iocore/net/quic/QUICAckFrameCreator.cc [hunk 2]
```diff
@@ -88,28 +75,204 @@ QUICAckFrameCreator::generate_frame(QUICEncryptionLevel level, uint64_t connecti
...
+
+  // unorder packet should send ack immediately to accellerate the recovery
+  if (this->_expect_next != packet_number) {
+    this->_expect_next = std::max(this->_expect_next, packet_number + 1);
+    this->_should_send = true;
+  }
+
+  // every 2 full-packet should send a ack frame like tcp
+  this->_size_unsend += size;
+  if (this->_size_unsend > 2 * 1480) {
```
Comment (scw00) [inline iocore/net/quic/QUICAckFrameCreator.cc]: should use PMTU

**Comment by maskit** [comment]
I know this is WIP and I just read this only for 1 min, but I have a couple of questions.

- common_send_packet()
Why is this needed? PACKET_WRITE_READY is scheduled like a spin now, and ideally the event should be scheduled only if the socket (buffer) is really writable.

- QUICAckPacketNumbers::_create_ack_frame
What's the motivation of adding this function?

**Comment by scw00** [comment]
Now it can be reviewed

**Comment by scw00** [comment]
>common_send_packet()
Why is this needed? PACKET_WRITE_READY is scheduled like a spin now, and ideally the event should be scheduled only if the socket (buffer) is really writable.

Because we need to packetize frame immediately, When delay's timer fired. It should signal QUICNetVConnection to send acks out.

>QUICAckPacketNumbers::_create_ack_frame 
What's the motivation of adding this function?

To cancel ack_frame which do not satisfy the given maximum_frame_size

**Comment by maskit** [comment]
> Because we need to packetize frame immediately, When delay's timer fired. It should signal QUICNetVConnection to send acks out.

I understand the ACK frame have to be sent with the minimal delay in this case, but how long will it be delayed if we didn't signal it and just waited the next event? In worst case, we would just waste one event loop due to socket buffer full.

I think an effective change would be calling `ack_frame_creator::generate_frame` as early as possible in `_packetize_frame`. I think it was originally called very early in the function. The reason QUICFrameGenerator has `will_generate_frame` is to check whether the creating packet will be ack_only packet, currently it's not used in many cases though.

**Comment by scw00** [comment]
>but how long will it be delayed if we didn't signal it and just waited the next event? In worst case, we would just waste one event loop due to socket buffer full.

Currently the PACKET_WRITE_READY_EVENT doesn't indicate the socket writable. 

>In worst case, we would just waste one event loop due to socket buffer full.

In worst case, that we can not write any byte to socket buffer, We can just discard packet because we are in a serious net storm. Peers will received another ack_frame (because missing ack is not strongly indicate the packet has lost) before marking the packet as lost.

>I think an effective change would be calling ack_frame_creator::generate_frame as early as possible in _packetize_frame.

That will wast bandwidth but it is ok, if you set `max_delay_ack = 0`. Then we always send ack immediately .

> The reason QUICFrameGenerator has will_generate_frame is to check whether the creating packet will be ack_only packet, currently it's not used in many cases though.

That's correct. So I still don't understand How will_generate_frame works. Can we just remove it ?

**Comment by scw00** [comment]
>QUIC SHOULD delay sending acknowledgements in response to packets, but MUST NOT excessively delay acknowledgements of retransmittable packets. Specifically, implementations MUST attempt to enforce a maximum ack delay to avoid causing the peer spurious timeouts. The maximum ack delay is communicated in the max_ack_delay transport parameter and the default value is 25ms.

**Comment by maskit** [comment]
> Currently the PACKET_WRITE_READY_EVENT doesn't indicate the socket writable.

This will be (has to be) changed eventually. The reason we currently do this is we have only one socket and all QUIC connection share it. But with connect+bind, we will be able to have one socket per connection like TCP connection. Then we can notify true PACKET_WRITE_READY_EVENT to each QNetVC.

> In worst case, that we can not write any byte to socket buffer, We can just discard packet because we are in a serious net storm. Peers will received another ack_frame (because missing ack is not strongly indicate the packet has lost) before marking the packet as lost.

I think we should avoid that scenario. If the packet has other frames, those frames will be lost too, and won't be retransmitted until they are regarded as lost.

> That will wast bandwidth but it is ok, if you set max_delay_ack = 0. Then we always send ack immediately .

No, bandwidth will not change at all. I'm not saying we should send ACK frame whenever it's available. I'm saying ACK frames should be stored to the head of packet payload. Currently we put ACK frame after STREAM frame, and because STREAM frames are usually big, they deprive ACK frame of a chance to be sent. If we put it at the head, we can ensure that ACK frame will be sent if AckFrameCreater think it has to be sent. Without this, ACK frame may not be sent even if we send the signal.

> That's correct. So I still don't understand How will_generate_frame works. Can we just remove it ?

We may be able remove it if we can guarantee that AckFrameCreator generates ACK frame only if we can really send it. As you know, we should not (or maybe must not) send back ack-only packet against ack-only packet.

The original design was like this (too difficult to explain in English..), and it was broken during changes.
```
bool will_be_ack_only = true;
for (generator : all_generators) {
  if (generator.will_generate_frame) {
    will_be_ack_only = false;
    break;
  }
}

if (will_be_ack_only) {
  if (ack_frame_creator.will_create_ack_frame_for_ack_only_packet) {
    // don't send the ack frame.    
  } else {
    // we can send the ack packet because it's not ack for ack-only packet
  }
} else {
  // we can send any ack frame even one for ack-only packet.
}

```

A reason I'm not sure we can remove `will_generate_frame` is that it seems like the function is used for other purpose in a few places. We need to check those use cases if we remove it.

**Comment by maskit** [comment]
>> What's the motivation of adding this function?

> To cancel ack_frame which do not satisfy the given maximum_frame_size

I meant `QUICAckPacketNumbers::create_ack_frame`. As I commented on the previous PR, I don't understand why you want to call `QUICFrameFactory::create_ack_frame` from `QUICAckPacketNumbers` but not `QUICAckFrameCreator`.

**Comment by scw00** [comment]
>To avoid creating an indefinite feedback loop, an endpoint MUST NOT send an ACK frame in response to a packet containing only ACK or PADDING frames, even if there are packet gaps which precede the received packet.

>As you know, we should not (or maybe must not) send back ack-only packet against ack-only packet.

That means Do not ack ack-only packet.  We can send ack-only packet to ack the packet  which is not ack-only even if the latest packet is ack-only.

**Comment by maskit** [comment]
> If you send a packet exceed the max_ack_delay it would mess the srtt and rtt . The recovery algro assume the acks are delayed for max_ack_delay time.

Does that mean we need to create ACK frame (at minimum) every <max_ack_delay> ms  even if we cannot send it (to reset internal timer)?

If so, AckFrameCreator need to ignore `connection_credit` and `maximum_frame_size`, right? And if I understand correctly, we should not packetize the ACK frame because we know we should not send it in terms of CC.

---

> Because every EncryptLevel has their own status and it should be checked individually. Of course we can do it together but it seems too complex and unclear .

Ah, now I understand the motivation. Let me think a bit more. The biggest reason I don't like the part is "Numbers" generate frames. I want numbers to be just data. Expressing theirselves is fine (e.g. toString() in Java) but creating frames is too much, I think. Maybe we can just rename it, or maybe we can create another structure.

---

> That means Do not ack ack-only packet. We can send ack-only packet to ack the packet which is not ack-only even if the latest packet is ack-only.

You're right, and the paragraph is what I wanted to say. But the logic of the pseudo code I wrote doesn't change. More accurate condition would be:

if (ack_frame_creator.will_create_ack_frame_ **just_for** _ack_only_packet) {`

**Comment by scw00** [comment]
>You're right, and the paragraph is what I wanted to say. But the logic of the pseudo code I wrote doesn't change. More accurate condition would be

So we can do the check in create_ack_frame 
See
```
 // skip ack_only packets
  for (; it != list.end(); it++) {
    if (!(*it).ack_only) {
      break;
    }
  }
   if (it == list.end()) {
    return ack_frame;
  }
```

Then we will never generate the ack which acknowledge  ack-only packet.

**Comment by scw00** [comment]
>Does that mean we need to create ACK frame (at minimum) every <max_ack_delay> ms even if we cannot send it (to reset internal timer)?

It is a good idea.

**Comment by maskit** [comment]
> So we can do the check in create_ack_frame

Seems like the spec confused us, because I just found a quite important difference between -15 and -16 while I was reading the code snippet you posted on the comment. There might be some more differences.

draft-15 says:
>  Implementations MUST NOT **generate packets that only contain ACK
   frames** in response to packets which only contain ACK and PADDING
   frames.

but draft-16 says:
> To avoid creating an indefinite feedback loop, an endpoint MUST NOT
   **send an ACK frame** in response to a packet containing only ACK or
   PADDING frames, even if there are packet gaps which precede the
   received packet.

If I understand these correctly, draft-15 allows sending ACK if it will be sent with other frames, but draft-16 doesn't.

However, according to the announcement of -16,  there should not be such important technical changes between -15 and -16.

> These drafts look a lot different to the -15 drafts, but there should
be no substantive technical changes.
 
I'll ask about this change on the slack channel.

File: iocore/net/quic/QUICAckFrameCreator.h [hunk 3]
```diff
@@ -23,47 +23,78 @@
 
 #pragma once
 
+#include "../../eventsystem/I_EventSystem.h"
```
Comment (maskit) [inline iocore/net/quic/QUICAckFrameCreator.h]: Just I_EventSystem.h seems fine. I changed other places.

**Comment by scw00** [comment]
Talked with @maskit on slack channel

There is the conclusion:
 - Do not schedule PACKET_READY_EVENT
 - Rename AckFrameCreator -> AckManager and AckPacketNumbers -> AckFrameCreator
 - Only acknowledge  packet  which is not ack-only. (The draft-15 can send ack+stream in response to ack-only packet but draft-16 not. See comment above).
 - Run QUICAckCreator as periodic event. 
 - Generate ack frame immediately just like Flow control. We can save the frame in somewhere then return it when someone call generate_frame. Because the ack frame is really time sensitive .

Any supply would be great appreciate .

File: iocore/net/QUICNetVConnection.cc [hunk 4]
```diff
@@ -197,6 +208,11 @@ QUICNetVConnection::acceptEvent(int event, Event *e)
     set_active_timeout(active_timeout_in);
   }
 
+  auto ethread = this_ethread();
+  if (ethread != nullptr) {
```
Comment (maskit) [inline iocore/net/QUICNetVConnection.cc]: Should we check this?

File: iocore/net/QUICNetVConnection.cc [hunk 5]
```diff
@@ -197,6 +208,11 @@ QUICNetVConnection::acceptEvent(int event, Event *e)
...
+    this->_ack_manager_periodic = ethread->schedule_every(this, HRTIME_MSECONDS(25), QUIC_EVENT_ACK_PERIODIC);
```
Comment (maskit) [inline iocore/net/QUICNetVConnection.cc]: Can we. move this into `start()`? `params->max_ack_delay()` is available.

File: iocore/net/quic/QUICAckFrameCreator.h (hunk 1 above)
Comment (maskit) [reply iocore/net/quic/QUICAckFrameCreator.h]: Not used now?

Comment (maskit) [reply iocore/net/quic/QUICAckFrameCreator.h]: Ah, cc file was not shown.

File: iocore/net/quic/QUICAckFrameCreator.cc (hunk 2 above)
Comment (maskit) [reply iocore/net/quic/QUICAckFrameCreator.cc]: Can you put FIXME comment for now?

File: iocore/net/QUICNetVConnection.cc [hunk 6]
```diff
@@ -599,6 +616,17 @@ QUICNetVConnection::state_handshake(int event, Event *data)
...
     break;
   }
+  case QUIC_EVENT_ACK_PERIODIC: {
+    if (!this->_refresh_ack_frame_manager()) {
+      break;
+    }
+
+    // we have ack to send
+    // FIXME: should sent depend on socket event.
+    this->_schedule_packet_write_ready();
```
Comment (scw00) [inline iocore/net/QUICNetVConnection.cc]: Changed here. To avoid the assertion in _close_packet_ready.

File: iocore/net/QUICNetVConnection.cc [hunk 7]
```diff
@@ -266,6 +277,12 @@ QUICNetVConnection::start()
   this->_frame_dispatcher->add_handler(this->_stream_manager);
   this->_frame_dispatcher->add_handler(this->_path_validator);
   this->_frame_dispatcher->add_handler(this->_handshake_handler);
+
+  if (this->thread) {
```
Comment (maskit) [inline iocore/net/QUICNetVConnection.cc]: Is this check for client side? I have no idea why some NetVC don't have any thread.

Comment (scw00) [reply iocore/net/QUICNetVConnection.cc]: On test ? or other case which event system not initialize completely.

Comment (maskit) [reply iocore/net/QUICNetVConnection.cc]: Hmm, I think we should not make such a situation happen. Behaving differently on tests doesn't make sense. And if it was not a test, we should definitely set a thread.
Did you find any actual cases that require this check while you are working on this?

Comment (scw00) [reply iocore/net/QUICNetVConnection.cc]: No. Ok I'll remove it and add assertion here.

**Comment by scw00** [comment]
updated!

**Pull request merged**